### ISSUE TITLE
Closing popup window UnhandledException

### ISSUE THREAD

**Issue by KennyKay** [cefsharp/CefSharp#1594]
I'm currently using WinForm CefSharp 47.0.2 and when I close a popup, there is a high chance of crashing the browser.

The exception message that I am receiving is:
ClientAdapter::CefSharp.Internals.ClientAdapter.OnResourceLoadComplete couldn't find IBrowser entry! Please contact CefSharp development.

Stacktrace:
```
   at CefSharp.Internals.ClientAdapter.GetBrowserWrapper(ClientAdapter* , Int32 browserId, Boolean isPopup)
   at CefSharp.Internals.ClientAdapter.OnResourceLoadComplete(ClientAdapter* , CefRefPtr<CefBrowser>* browser, CefRefPtr<CefFrame>* frame, CefRefPtr<CefRequest>* request, CefRefPtr<CefResponse>* response, cef_urlrequest_status_t status, Int64 receivedContentLength)
```

Thanks,

**Comment by turgayselvi** [comment]
I have noticed same issue CefSharp 43 also. It doesn't crash for every page in a popup. 

To reproduce the issue:

1- ) Go to: www.yahoo.com
2-) Type 'NBA' on search panel.
3-) Click NBA.com on the list. It will open as popup.
4-) Close popup window. 
5-) It will crash.

After popup browser is deleted from browser dictionary. GetBrowserWrapper is called. This causes an issue. But not for every page. That is strange.


ClientAdapter.cpp


   IBrowser^ ClientAdapter::GetBrowserWrapper(int browserId, bool isPopup)
        {
            if (_browserControl->HasParent)
            {
                return _browserAdapter->GetBrowser();
            }

            if(isPopup)
            {
                IBrowser^ browserWrapper;
                if (_popupBrowsers->TryGetValue(browserId, browserWrapper))
                {
                    return browserWrapper;
                }

                auto stackFrame = gcnew StackFrame(1);
                auto callingMethodName = stackFrame->GetMethod()->Name;

            **    ThrowUnknownPopupBrowser(gcnew String(L"ClientAdapter::" + callingMethodName));**
            }

            return _browserAdapter->GetBrowser();
        }

**Comment by amaitland** [comment]
It appears to be a timing issue, would depend somewhat on the web page and network connection. If you close the popup before all it's resources have completed downloading then an exception will be thrown.

**Comment by amaitland** [comment]
Commit https://github.com/cefsharp/CefSharp/commit/87461203b566a1b464c90c250acd78b4b1c959b8 should resolve this.

If someone can perform additional testing and confirm that would be appreciated. Once this is done I'll look at releasing `47.0.3`.

**Comment by turgayselvi** [comment]
Hi Alex,

I have tested it. It is till crashing.

**returnNullIfBrowserNotFound** field is always false.



        IBrowser^ ClientAdapter::GetBrowserWrapper(int browserId, bool isPopup, bool returnNullIfBrowserNotFound)
        {
            if (_browserControl->HasParent)
            {
                return _browser;
            }

            if(isPopup)
            {
                IBrowser^ popupBrowser;
                if (_popupBrowsers->TryGetValue(browserId, popupBrowser))
                {
                    return popupBrowser;
                }

                if (returnNullIfBrowserNotFound)
                {
                    return nullptr;
                }

                auto stackFrame = gcnew StackFrame(1);
                auto callingMethodName = stackFrame->GetMethod()->Name;

                ThrowUnknownPopupBrowser(gcnew String(L"ClientAdapter::" + callingMethodName));
            }

**Comment by turgayselvi** [comment]
You set returnNullIfBrowserNotFound as true in **ClientAdapter::OnResourceLoadComplete.**

It is called **cef_return_value_t ClientAdapter::OnBeforeResourceLoad **after click close popup.  It send "false".  This causes an exception.

**Comment by amaitland** [comment]
@turgayselvi The original post says `OnResourceLoadComplete` and your follow up doesn't provide a stack trace, so your problem is actually different to the one outlined.

You can submit a `PR` with your proposed change.

**Comment by turgayselvi** [comment]
@amaitland I was getting same error before that is in stack trace. But after get latest update, Stack trace points "OnBeforeResourceLoad" .  

I set returnNullIfBrowserNotFound field value as "True" for **GetBrowserWrapper** method in **OnBeforeResourceLoad**.  Now it works without crashing.


        cef_return_value_t ClientAdapter::OnBeforeResourceLoad(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame, CefRefPtr<CefRequest> request, CefRefPtr<CefRequestCallback> callback)
        {
            auto handler = _browserControl->RequestHandler;

            if (handler == nullptr)
            {
                return cef_return_value_t::RV_CONTINUE;
            }            

            auto frameWrapper = gcnew CefFrameWrapper(frame);
            auto browserWrapper = GetBrowserWrapper(browser->GetIdentifier(), browser->IsPopup(), true);
            auto requestWrapper = gcnew CefRequestWrapper(request);
            auto requestCallback = gcnew CefRequestCallbackWrapper(callback, frameWrapper, requestWrapper);

            return (cef_return_value_t)handler->OnBeforeResourceLoad(_browserControl, browserWrapper, frameWrapper, requestWrapper, requestCallback);
        }

**Comment by amaitland** [comment]
> I set returnNullIfBrowserNotFound field value as "True" for GetBrowserWrapper method in OnBeforeResourceLoad. Now it works without crashing.

Great :+1: Please submit a `PR` and I'll merge it later today.

**Comment by amaitland** [comment]
Changes merged into `cefsharp/47` branch.

Will release when I have time to perform some additional testing.

**Comment by amaitland** [comment]
The `47.0.3` release has been tagged, will appear on the `MyGet` feed shortly. If all goes well I'll push it to `Nuget.org` this afternoon/tomorrow.

**Comment by amaitland** [comment]
Release is on `Nuget.org`

**Comment by edgchen1** [comment]
@amaitland @turgayselvi 

Hi,
Sometimes I still see an ApplicationException when closing a popup window.
```
System.ApplicationException was unhandled
Message: An unhandled exception of type 'System.ApplicationException' occurred in CefSharp.Core.dll
Additional information: ClientAdapter::CefSharp.Internals.ClientAdapter.OnResourceResponse couldn't find IBrowser entry! Please contact CefSharp development.
```

There are 2 stack traces that I've seen when this occurs:
```
	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::ThrowUnknownPopupBrowser(System::String^ context) Line 48	C++
 	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::GetBrowserWrapper(int browserId, bool isPopup, bool returnNullIfBrowserNotFound) Line 86	C++
 	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::GetResourceHandler(CefRefPtr<CefResourceHandler>* this, CefRefPtr<CefBrowser>* browser, CefRefPtr<CefFrame>* frame, CefRefPtr<CefRequest>* request) Line 663	C++
```
```
 	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::ThrowUnknownPopupBrowser(System::String^ context) Line 48	C++
 	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::GetBrowserWrapper(int browserId, bool isPopup, bool returnNullIfBrowserNotFound) Line 86	C++
	CefSharp.Core.dll!CefSharp::Internals::ClientAdapter::OnResourceResponse(CefRefPtr<CefBrowser>* browser, CefRefPtr<CefFrame>* frame, CefRefPtr<CefRequest>* request, CefRefPtr<CefResponse>* response) Line 615	C++
```

I am running the CefSharp.Wpf.Example program in CefSharp 47.0.3.

Steps to reproduce:
- Navigate to www.yahoo.com
- Search for "NBA"
- Click the link to "NBA.com" multiple times (e.g. 10 times)
- Close the resulting popup windows fairly quickly (e.g., less than one second between closes)
- The exception is thrown.

Note that the exception does not reliably occur at same point - sometimes I need to close more popups than other times.

**Comment by amaitland** [comment]
@edgchen1 For the call to `GetBrowserWrapper` flip the last boolean, If that works successfully then please submit a `PR` :+1:
